src/prefect/results.py

```diff
@@ -360,20 +369,194 @@ def key_fn():
+        except Exception as exc:
+            extra_info = (
+                'You can try a different serializer (e.g. result_serializer="json") '
+                "or disabling persistence (persist_result=False) for this flow or task."
+            )
+            # check if this is a known issue with cloudpickle and pydantic
+            # and add extra information to help the user recover
+
+            if (
+                isinstance(exc, TypeError)
+                and isinstance(value, BaseModel)
+                and str(exc).startswith("cannot pickle")
+            ):
+                try:
+                    from IPython import get_ipython
+
+                    if get_ipython() is not None:
+                        extra_info = inspect.cleandoc(
+                            """
+                            This is a known issue in Pydantic that prevents
+                            locally-defined (non-imported) models from being
+                            serialized by cloudpickle in IPython/Jupyter
+                            environments. Please see
+                            https://github.com/pydantic/pydantic/issues/8232 for
+                            more information. To fix the issue, either: (1) move
+                            your Pydantic class definition to an importable
+                            location, (2) use the JSON serializer for your flow
+                            or task (`result_serializer="json"`), or (3)
+                            disable result persistence for your flow or task
+                            (`persist_result=False`).
+                            """
+                        ).replace("\n", " ")
+                except ImportError:
+                    pass
+            raise ValueError(
+                f"Failed to serialize object of type {type(value).__name__!r} with "
+                f"serializer {self.serializer.type!r}. {extra_info}"
+            ) from exc
+
+        return data
+
+    @model_validator(mode="before")
+    @classmethod
+    def coerce_old_format(cls, value: Any):
+        if isinstance(value, dict):
+            if "data" in value:
+                value["result"] = value.pop("data")
+            if "metadata" not in value:
+                value["metadata"] = {}
+            if "expiration" in value:
+                value["metadata"]["expiration"] = value.pop("expiration")
+            if "serializer" in value:
+                value["metadata"]["serializer"] = value.pop("serializer")
+            if "prefect_version" in value:
+                value["metadata"]["prefect_version"] = value.pop("prefect_version")
+        return value
+
+    def serialize_metadata(self) -> bytes:
+        return self.metadata.dump_bytes()
+
+    def serialize(
+        self,
+    ) -> bytes:
+        """
+        Serialize the record to bytes.
+
+        Either or both of `include_metadata` and `include_result` must be True.
```

Comment:
what does this mean?

Reply:
`include_metadata` and `include_result` can't both be `False`. This could also be handled by separate methods.

Reply:
Right but what are they? Based on the docstring I'd expect them to be kwargs to this method but I don't see any other references. Maybe an oversight?

Reply:
Oh yeah, I must've accidentally taken those out. I'll bring them back.

Reply:
otherwise LGTM

Reply:
Alright, I remember now, there's `serialize`, `serialize_result`, and `serialize_metadata` for when the entire record needs to be serialized and when the metadata and results need to be serialized separately. This docstring was left over from when I changed from one method with kwargs to three methods without kwargs. I've removed the dangling docstring.

Reply:
Sounds good

Reply:
Approved - for my curiosity, why'd you move from kwargs to distinct methods @desertaxle ?

Reply:
My reasoning was that the three valid modes didn't share any logic, and having three separate methods eliminated a potential foot gun.